### PR TITLE
Streamlines the usage of `console-ui`

### PR DESCRIPTION
**Pull request by twokul**
Introduces a new model to serve as a singleton for `console-ui`.

cc @rwjblue

### CODE REVIEW

File: lib/models/project.js [hunk 1]
```diff
@@ -625,14 +620,18 @@ Project.closestSync = function(pathName, _ui, _cli) {
   @static
   @method projectOrnullProject
   @param  {UI} _ui The UI instance to provide to the created Project.
+  @param  {CLI} _cli The CLI instance to provide to the created Project.
   @return {Project}         Project instance
  */
-Project.projectOrnullProject = function(_ui, _cli) {
+Project.projectOrnullProject = function(ui, cli) {
+  this.ui = ui;
```
Comment (rwjblue) [inline lib/models/project.js]: This seems _slightly_ odd here, because `this` is actually `Project.prototype` here.

Comment (twokul) [inline lib/models/project.js]: yeah I know

Comment (stefanpenner) [inline lib/models/project.js]: `this` is actually `Project` itself here, not `Project.prototype`. We should not put `state` such as `ui` on the `Project` class itself.

Comment (twokul) [inline lib/models/project.js]: agreed, I was working on this over the weekend, decided to go with a singleton + a separate module; only 2 tests are failing, need to fix before pushing

File: lib/models/project.js [hunk 2]
```diff
@@ -575,20 +576,22 @@ class Project {
     @static
     @method closest
     @param  {String} pathName Path to your project
+    @param  {UI} optionalUI The UI instance to provide to the created Project.
+    @param  {Object} optionalCLI The CLI instnace to provide to the created Project.
```
Comment (btecu) [inline lib/models/project.js]: `instnace` -> `instance`

File: lib/models/project.js [hunk 3]
```diff
@@ -600,32 +603,33 @@ class Project {
     @static
     @method closestSync
     @param  {String} pathName Path to your project
-    @param  {UI} _ui The UI instance to provide to the created Project.
+    @param  {UI} optionalUI The UI instance to provide to the created Project.
+    @param  {Object} optionalCLI The CLI instnace to provide to the created Project.
```
Comment (btecu) [inline lib/models/project.js]: `instnace` ...

File: lib/models/project.js [hunk 4]
```diff
@@ -640,12 +644,12 @@ class Project {
     @param  {UI} _ui The UI instance to provide to the created Project.
```
Comment (nathanhammond) [inline lib/models/project.js]: Should fix documentation here as well.

File: lib/models/ui.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  constructor(args) {
+    let options = args || {};
+
+    if (!singletonUI) {
+      singletonUI = new UI({
+        inputStream: options.inputStream,
+        outputStream: options.outputStream,
+        errorStream: options.errorStream || process.stderr,
+        errorLog: options.errorLog || [],
+        ci: process.env.CI || (/^(dumb|emacs)$/).test(process.env.TERM),
```
Comment (nathanhammond) [inline lib/models/ui.js]: I know this is just moved, I'm unsure why `dumb`, though?

File: lib/models/ui.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * The class that serves as a singleton wrapper for `console-ui`.
+ *
+ * @private
+ * @module ember-cli
+ * @class CommandLineUI
+ * @constructor
+ */
+class CommandLineUI {
```
Comment (nathanhammond) [inline lib/models/ui.js]: Oh how I love this new year.

File: lib/models/ui.js (hunk 5 above)
Comment (twokul) [inline lib/models/ui.js]: I believe if terminal has no ANSI support then `TERM` will be set to `dumb` (meaning there will be no colours); it is sort of a "last resort, when everything is broken" kind of thing.

File: lib/models/ui.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  constructor(args) {
+    let options = args || {};
+
+    if (!singletonUI) {
+      singletonUI = new UI({
+        inputStream: options.inputStream,
```
Comment (kellyselden) [inline lib/models/ui.js]: Do these need to be `inputStream: options.inputStream || process.stdin,` to support this:
https://github.com/ember-cli/ember-cli/pull/6577/files#diff-3677132bf6240059b3a3637716019c84L711

Comment (nathanhammond) [inline lib/models/ui.js]: I think so.

Comment (twokul) [inline lib/models/ui.js]: I don't think so. We instantiate UI only once in `lib/cli/index.js` and `process.stdin` is passed via `options` from [`bin/ember`](https://github.com/ember-cli/ember-cli/blob/master/bin/ember#L44); in tests, we use a fake UI from `console-ui`.

Comment (kellyselden) [inline lib/models/ui.js]: In this location, https://github.com/ember-cli/ember-cli/pull/6577/files#diff-3677132bf6240059b3a3637716019c84R613, it used to use `process` in the fallback, but now it uses null options.

Comment (twokul) [inline lib/models/ui.js]: we use `closestSync` only in one place - [`lib/models/project`](https://github.com/ember-cli/ember-cli/search?utf8=%E2%9C%93&q=closestSync) and we do not pass in `ui` or `cli` parameters when we call it.

the reason the fallback was put in place was b/c we created 2 instances of the `UI`: one when `bin/ember` runs and the other one when we require `lib/broccoli/ember-app`.

with this change, `UI` is created only once and we no longer require this fallback.

perhaps I'm missing something obvious?

Comment (nathanhammond) [inline lib/models/ui.js]: Nope, it seems like you know this code better than the both of us. 😛

**Comment by Turbo87** [comment]
@twokul can you rebased this? we should try to push this over the finish line

**Comment by twokul** [comment]
@Turbo87 sure thing. let me rebase

**Comment by twokul** [comment]
@Turbo87 done

**Comment by locks** [comment]
Closing due to inactivity.

**Comment by locks** [comment]
Closing due to inactivity.

**Pull request closed without merging**